payjoin/src/receive/multi_party/mod.rs

```diff
@@ -0,0 +1,253 @@
+use bitcoin::{FeeRate, Psbt};
+use error::{InternalMultiPartyError, MultiPartyError};
+
+use super::error::InputContributionError;
+use super::{v1, v2, Error, ImplementationError, InputPair};
+use crate::psbt::merge::merge_unsigned_tx;
+use crate::receive::v2::SessionContext;
+
+pub(crate) mod error;
+
+const SUPPORTED_VERSIONS: &[usize] = &[2];
+
+#[derive(Default)]
+pub struct UncheckedProposalBuilder {
+    proposals: Vec<v2::UncheckedProposal>,
+}
+
+impl UncheckedProposalBuilder {
+    pub fn new() -> Self { Self { proposals: vec![] } }
+
+    pub fn try_add(
+        &mut self,
+        proposal: v2::UncheckedProposal,
+    ) -> Result<&mut Self, MultiPartyError> {
+        self.check_proposal_suitability(&proposal)?;
+        self.proposals.push(proposal);
+        Ok(self)
+    }
+
+    fn check_proposal_suitability(
+        &self,
+        proposal: &v2::UncheckedProposal,
+    ) -> Result<(), MultiPartyError> {
+        let params = proposal.v1.params.clone();
+        if !SUPPORTED_VERSIONS.contains(&params.v) {
+            return Err(InternalMultiPartyError::ProposalVersionNotSupported(params.v).into());
+        }
+
+        if !params.optimistic_merge {
+            return Err(InternalMultiPartyError::OptimisticMergeNotSupported.into());
+        }
+        Ok(())
+    }
+
+    pub fn try_build(&self) -> Result<UncheckedProposal, MultiPartyError> {
```

Comment:
`try_`, again, is for fallable functions. This is `build()`.

Reply:
done